toolchain/check/subst.cpp

```diff
@@ -0,0 +1,241 @@
+
+  // Note that we pop in reverse order because we pushed them in forwards order.
+  int32_t arg1 = PopOperand(context, worklist, kinds.second, inst.arg1());
+  int32_t arg0 = PopOperand(context, worklist, kinds.first, inst.arg0());
+  if (arg0 == inst.arg0() && arg1 == inst.arg1()) {
+    return inst_id;
+  }
+
+  inst.SetArgs(arg0, arg1);
+  auto result_id = TryEvalInst(context, SemIR::InstId::Invalid, inst);
+  CARBON_CHECK(result_id.is_constant())
+      << "Substitution into constant produced non-constant";
+  return result_id.inst_id();
+}
+
+auto SubstConstant(Context& context, SemIR::ConstantId const_id,
+                   Substitutions substitutions) -> SemIR::ConstantId {
+  CARBON_CHECK(const_id.is_constant()) << "Substituting into non-constant";
+
+  if (substitutions.empty()) {
+    // Nothing to substitute.
+    return const_id;
+  }
+
+  if (!const_id.is_symbolic()) {
+    // A template constant can't contain a reference to a symbolic binding.
+    return const_id;
+  }
+
+  Worklist worklist;
+  worklist.Push(const_id.inst_id());
+  worklist.Get(0).next_index = -1;
+
+  // For each instruction that forms part of the constant, we will visit it
+  // twice:
+  //
+  // - First, we visit it with `is_expanded == false`, we add all of its
+  //   operands onto the worklist, and process them by following this same
+  //   process.
+  // - Then, once all operands are processed, we visit the instruction with
+  //   `is_expanded == true`, pop the operands back off the worklist, and if any
+  //   of them changed, rebuild this instruction.
+  //
+  // The second step is skipped if we can detect in the first step that the
+  // instruction will not need to be rebuilt.
+  int index = 0;
+  while (index != -1) {
+    auto& item = worklist.Get(index);
+
+    if (item.is_expanded) {
+      // Rebuild this item if necessary. Note that this might pop items from the
+      // worklist but does not reallocate, so does not invalidate `item`.
+      item.inst_id = Rebuild(context, worklist, item.inst_id);
+      index = item.next_index;
+      continue;
+    }
+
+    if (context.constant_values().Get(item.inst_id).is_template()) {
+      // This instruction is a template constant, so can't contain any
+      // bindings that need to be substituted.
+      index = item.next_index;
+      continue;
+    }
+
+    if (context.insts().Is<SemIR::BindSymbolicName>(item.inst_id)) {
+      // This is a symbolic binding. Check if we're substituting it.
+
+      // TODO: Consider building a hash map for substitutions. We might have a
+      // lot of them.
+      for (auto [bind_id, replacement_id] : substitutions) {
+        if (item.inst_id == bind_id) {
+          // This is the binding we're replacing. Perform substitution.
+          item.inst_id = replacement_id.inst_id();
+          break;
+        }
+      }
+
+      // Even if it's not being substituted, don't look through it. Its constant
+      // value doesn't spend on its operand.
```

Comment:
```suggestion
      // If it's not being substituted, don't look through it. Its constant
      // value doesn't depend on its operand.
```

Minor phrasing suggestion, feel free to reject if I'm misunderstanding, although I think "spend" is a typo.